src/ring_theory/ideal/operations.lean

```diff
@@ -1051,6 +1051,38 @@ by { rw [mem_ker, f.map_one], exact one_ne_zero }
 @[simp] lemma ker_coe_equiv (f : R ≃+* S) : ker (f : R →+* S) = ⊥ :=
 by simpa only [←injective_iff_ker_eq_bot] using f.injective
 
+/-- The induced map from the quotient by the kernel to the codomain. -/
+def ker_lift (f : R →+* S) : f.ker.quotient →+* S :=
+ideal.quotient.lift _ f $ λ r, f.mem_ker.mp
+
+@[simp]
+lemma ker_lift_mk (f : R →+* S) (r : R) : (ker_lift f) (ideal.quotient.mk f.ker r) = f r :=
+ideal.quotient.lift_mk _ _ _
+
+/-- The induced map from the quotient by the kernel is injective. -/
+lemma injective_ker_lift (f : R →+* S) : function.injective (ker_lift f) :=
+assume a b, quotient.induction_on₂' a b $
+  assume a b (h : f a = f b), quotient.sound' $
+show a - b ∈ ker f, by rw [mem_ker, map_sub, h, sub_self]
+
+variable {f}
+
+/-- If `f : R →+* S` is surjective, then the induced map from the quotient by the kernel is
+surjective. -/
+lemma surjective_ker_lift (hf : function.surjective f) : function.surjective (ker_lift f) :=
+begin
+  intro s,
+  rcases hf s with ⟨r, rfl⟩,
+  refine ⟨ideal.quotient.mk _ r, rfl⟩,
+end
+
+/-- The first isomorphism theorem for commutative rings. -/
+noncomputable def quotient_ker_equiv_of_surjective (hf : function.surjective f) :
+  f.ker.quotient ≃+* S :=
+ring_equiv.of_bijective (ker_lift f) $
+{ left := injective_ker_lift f,
+  right := surjective_ker_lift hf }
```

Comment:
You can make this computable as
```suggestion
/-- The first isomorphism theorem for commutative rings. -/
def quotient_ker_equiv_of_right_inverse
  {g : S →+* R} (hf : function.right_inverse g f) :
  f.ker.quotient ≃+* S :=
{ to_fun := ker_lift f,
  inv_fun := (ideal.quotient.mk f.ker).comp g,
  left_inv := begin
    rintro ⟨x⟩,
    apply injective_ker_lift,
    simp [hf (f x)],
  end,
  right_inv := hf,
  ..ker_lift f}
```
I assume there's a lemma somewhere that constructs `g` from `function.surjective f` noncomputably so you can write `of_surjective` in terms of `of_right_inverse`?

Reply:
There are no morphism from `Z/2Z` to `Z` :)

Reply:
There is an interesting post on mathoverflow that explains why it is reasonable to bother about the axiom of choice, in the form "every surjection admits a section" (even if we don't care at all about the other axioms): in categories other than set it is false. I found this point of view very interesting.

Reply:
Whoops, here's the correct version without that mistake, where `g` is not a morphism!

```suggestion
/-- The first isomorphism theorem for commutative rings. -/
def quotient_ker_equiv_of_right_inverse
  {g : S → R} (hf : function.right_inverse g f) :
  f.ker.quotient ≃+* S :=
{ to_fun := ker_lift f,
  inv_fun := (ideal.quotient.mk f.ker) ∘ g,
  left_inv := begin
    rintro ⟨x⟩,
    apply injective_ker_lift,
    simp [hf (f x)],
  end,
  right_inv := hf,
  ..ker_lift f}

/-- The first isomorphism theorem for commutative rings. -/
noncomputable def quotient_ker_equiv_of_surjective (hf : function.surjective f) :
  f.ker.quotient ≃+* S :=
quotient_ker_equiv_of_right_inverse (classical.some_spec hf.has_right_inverse)
```

Reply:
(which removes the need for `surjective_ker_lift`)

Reply:
That makes sense! Time to sleep for me, tomorrow I will modify the algebra part too.